### PR TITLE
Top-Pruning: Top-Pruning Engine & Init on Tree

### LINKED ISSUE #453: Top-Pruning: Top-Pruning Engine & 2-Pass
This is an analog to Generalized Pruning's `GPEngine`, Top Pruning's `TPEngine` will be an object for holding a `PVHandler`, `ChoiceMap`, branch length vector, and other DAG associated data needed for performing Top Pruning, as well as methods for doing so.  This will be used by the NNIEngine in order to perform a systematic NNI search using Top Pruning.  

This issue will include performing an initialization step that finds a `ChoiceMap` for the best tree for all edges in the DAG.  The initialization step will include a 2-pass algorithm over the DAG (described in issue #440, part 3).  Steps 4 and beyond will be left to future issues.

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

TPEngine
- Outlines API for TPEngine.
- Supports resizing parameters to fit DAG and GraftDAG.
- Functionality for initializing branch lengths by taking first occurrance.
- Functionality for initializing choice map for simple tree.

Closes #453 

## Tests

TEST_CASE("Top-Pruning: TPEngine"): Compare the tree found by TPEngine to a TreeCollection built from the DAG, assures that all topologies created from TPEngine exist in DAG.

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Issue branch has been squashed and rebased on main branch
* [x] GitHub CI build on PR branch completed successfully

### CODE REVIEW

**Review by matsen — COMMENTED** [review]
A few minor requests here.

File: src/gp_doctest.cpp [hunk 1]
```diff
@@ -1757,3 +1757,35 @@ TEST_CASE("Top-Pruning: ChoiceMap") {
                   "Edge resulted in an invalid Topology not contained in DAG.");
   }
 }
+
+// This test uses the TPEngine to map of the top trees. Initializes the TPEngine choice
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_doctest.cpp]: Cleaned up test case language.

File: src/tp_engine.hpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  void InitializeChoiceMap();
+  // Get top-scoring tree in DAG containing given edge.
+  Node::NodePtr GetTopTreeTopologyWithEdge(const size_t edge_id) const;
+  // Get score of top-scoring tree in DAG containing given edge.
+  double GetScoreOfTopTree(const size_t edge_id) const;
+
+  // ** Scoring by Likelihood
+
+  // Initialize ChoiceMap for entire DAG by Likelihood, using 2-Pass Algorithm.
+  void InitializeDAGByLikelihood();
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: Changed name.

File: src/tp_engine.hpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  // Initialize ChoiceMap for entire DAG by Parsimony.
+  void InitializeDAGByParsimony();
+  // Update ChoiceMap with NNI.
+  void UpdateDAGAfterAddNodePairByParsimony(const NNIOperation &nni_op);
+
+  // ** Parameters
+
+  // Resize GPEngine to accomodate DAG with given number of nodes and edges.  Option to
+  // remap data according to DAG reindexers.  Option to give explicit numver of nodes or
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: Fixed typo.

File: src/tp_engine.hpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+#include "sugar.hpp"
+#include "gp_dag.hpp"
+#include "graft_dag.hpp"
+#include "pv_handler.hpp"
+#include "choice_map.hpp"
+#include "nni_operation.hpp"
+
+class TPEngine {
+ public:
+  using DAGNodeParameter = EigenVectorXd;
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: Removed alias.

File: src/tp_engine.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    const RootedTreeCollection &tree_collection, const BitsetSizeMap &edge_indexer) {
+  size_t unique_edge_count = branch_lengths_.size();
+  branch_lengths_.setZero();
+  EigenVectorXi observed_edge_counts = EigenVectorXi::Zero(unique_edge_count);
+  [&observed_edge_counts, this](size_t edge_idx, const RootedTree &tree,
+                                const Node *focal_node) {
+    branch_lengths_(edge_idx) += tree.BranchLength(focal_node);
+    observed_edge_counts(edge_idx)++;
+  };
+  // FunctionOverRootedTreeCollection(tally_branch_lengths_and_edge_count,
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.cpp]: Removed.

File: src/choice_map.hpp [hunk 6]
```diff
@@ -21,13 +21,31 @@ class ChoiceMap {
...
   ChoiceMap(GPDAG &dag)
       : dag_(dag), edge_choice_vector_(dag.EdgeCountWithLeafSubsplits()){};
 
+  // ** Access
+
+  const EdgeChoice &GetEdgeChoice(const size_t edge_id) const {
+    return edge_choice_vector_[edge_id];
+  }
+
+  void SetBestParentOfEdge(const size_t edge_id, const size_t parent_edge_id) {
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/choice_map.hpp]: Added enum and combined to single function.

File: src/tp_engine.hpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  // Absolute lower bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double min_log_branch_length_ = -13.9;
+  // Absolute upper bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double max_log_branch_length_ = 1.1;
+
+  // ** Scoring
+  // Partial Vector for storing Likelihoods or Parsimony scores.
+  PSVHandler pv_handler_;
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: Separated handlers.

File: src/tp_engine.hpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
+// Copyright 2019-2022 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+//
+// TPEngine runs the Top-Pruning algorithm.  Top-Pruning is a method for systematic
+// exploration, which finds .
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: Finished description.

File: src/tp_engine.hpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+  void SetEdgeCount(const size_t edge_count) { edge_count_ = edge_count; }
+  void SetSpareEdgeCount(const size_t edge_spare_count) {
+    edge_spare_count_ = edge_spare_count;
+  }
+  void SetAllocatedEdgeCount(const size_t edge_alloc) { edge_alloc_ = edge_alloc; }
+
+  // ** Access
+
+  NucleotidePLVRefVector &GetPVs() { return pv_handler_.GetPVs(); }
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp]: The PVHandlers store `NucleotidePLVRefVector`, but we could generalize this.  But I'd suggest saving it for a future issue.

File: src/gp_doctest.cpp [hunk 10]
```diff
@@ -1757,3 +1757,35 @@ TEST_CASE("Top-Pruning: ChoiceMap") {
...
+TEST_CASE("Top-Pruning: TPEngine Simple Tree") {
+  // Simple DAG that contains a single tree.
+  const std::string fasta_path = "data/six_taxon.fasta";
+  const std::string newick_path = "data/six_taxon_rooted_simple.nwk";
+  // Build DAG
+  auto inst = GPInstanceOfFiles(fasta_path, newick_path);
+  GPDAG& dag = inst.GetDAG();
+  inst.EstimateBranchLengths(0.00001, 100, true);
+  auto all_trees = inst.GenerateCompleteRootedTreeCollection();
+  TPEngine tpengine = TPEngine(dag, 8, "_ignore/mmapped_plv.data");
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_doctest.cpp]: Took SitePattern from inst.

File: src/tp_engine.hpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  // space).
+  static constexpr double min_log_branch_length_ = -13.9;
+  // Absolute upper bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double max_log_branch_length_ = 1.1;
+
+  // ** Scoring
+  // Partial Vector for storing Likelihood scores.
+  bool using_likelihoods_;
+  PSVHandler pv_likelihoods_;
```
Comment (matsen) [inline src/tp_engine.hpp:161]: Do you think `likelihood_pvs` is more appropriate? (ditto below)

File: src/tp_engine.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+  // !ISSUE #440
+  Failwith("Currently no implementation.");
+}
+
+void TPEngine::UpdateDAGAfterAddNodePairByParsimony(const NNIOperation &nni_op) {
+  // !ISSUE #440
+  Failwith("Currently no implementation.");
+}
+
+// ** Parametersz
```
Comment (matsen) [inline src/tp_engine.cpp:63]:
```suggestion
// ** Parameters
```

**Review by matsen — APPROVED** [review]
Great! Approved but please fix a few little things here.

File: src/tp_engine.hpp (hunk 11 above)
Comment (davidrich27) [reply src/tp_engine.hpp]: Yes. Name changed.

File: src/tp_engine.hpp [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  static constexpr double default_branch_length_ = 0.1;
+  // Absolute lower bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double min_log_branch_length_ = -13.9;
+  // Absolute upper bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double max_log_branch_length_ = 1.1;
+
+  // ** Scoring
+  // Partial Vector for storing Likelihood scores.
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/tp_engine.hpp:159]: Moved comment.

File: src/gp_doctest.cpp [hunk 14]
```diff
@@ -1757,3 +1757,37 @@ TEST_CASE("Top-Pruning: ChoiceMap") {
...
+
+// Initializes the TPEngine choice map and retrieves the top tree for each edge in DAG.
+// Then finds all trees contained in the DAG and verifies that each top tree produced is
+// a tree from the DAG.
+TEST_CASE("Top-Pruning: Initialize TPEngine and ChoiceMap") {
+  const std::string fasta_path = "data/six_taxon.fasta";
+  const std::string newick_path = "data/six_taxon_rooted_simple.nwk";
+
+  auto inst = GPInstanceOfFiles(fasta_path, newick_path);
+  // GPEngine& gpengine = *inst.GetEngine();
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_doctest.cpp]: Yes, removed.

File: src/tp_engine.cpp (hunk 12 above)
Comment (davidrich27) [reply src/tp_engine.cpp]: Fixed.

**Pull request merged**